clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1001,311 +1002,580 @@ static ParamDesc makeParamDesc(const FieldDecl *Src, QualType Ty) {
+  (void)std::initializer_list<int> { (handlers.FUNC(Field, FieldTy), 0)... }
+
+  for (const auto &Field : Fields) {
+    QualType FieldTy = Field->getType();
+
+    if (Util::isSyclAccessorType(FieldTy))
+      KF_FOR_EACH(handleSyclAccessorType);
+    else if (Util::isSyclSamplerType(FieldTy))
+      KF_FOR_EACH(handleSyclSamplerType);
+    else if (Util::isSyclSpecConstantType(FieldTy))
+      KF_FOR_EACH(handleSyclSpecConstantType);
+    else if (Util::isSyclStreamType(FieldTy))
+      KF_FOR_EACH(handleSyclStreamType);
+    else if (FieldTy->isStructureOrClassType()) {
+      KF_FOR_EACH(handleStructType);
+      CXXRecordDecl *RD = FieldTy->getAsCXXRecordDecl();
+      VisitAccessorWrapper(nullptr, Field, RD, handlers...);
+    } else if (FieldTy->isReferenceType())
+      KF_FOR_EACH(handleReferenceType);
+    else if (FieldTy->isPointerType())
+      KF_FOR_EACH(handlePointerType);
+    else if (FieldTy->isArrayType())
+      KF_FOR_EACH(handleArrayType);
+    else if (FieldTy->isScalarType())
+      KF_FOR_EACH(handleScalarType);
+    else
+      KF_FOR_EACH(handleOtherType);
+  }
+#undef KF_FOR_EACH
+}
+
+// A base type that the SYCL OpenCL Kernel construction task uses to implement
+// individual tasks.
+template <typename Derived> class SyclKernelFieldHandler {
+protected:
+  Sema &SemaRef;
+  SyclKernelFieldHandler(Sema &S) : SemaRef(S) {}
+
+public:
+  // Mark these virutal so that we can use override in the implementer classes,
+  // despite virtual dispatch never being used.
+
+  //// TODO: Can these return 'bool' and we can short-circuit the handling? That
+  // way the field checker cna return true/false based on whether the rest
+  // should be still working.
+
+  // Accessor can be a base class or a field decl, so both must be handled.
+  virtual void handleSyclAccessorType(const CXXBaseSpecifier &, QualType) {}
+  virtual void handleSyclAccessorType(const FieldDecl *, QualType) {}
+  virtual void handleSyclSamplerType(const FieldDecl *, QualType) {}
+  virtual void handleSyclSpecConstantType(const FieldDecl *, QualType) {}
+  virtual void handleSyclStreamType(const CXXBaseSpecifier &, QualType) {}
+  virtual void handleSyclStreamType(const FieldDecl *, QualType) {}
+  virtual void handleStructType(const FieldDecl *, QualType) {}
+  virtual void handleReferenceType(const FieldDecl *, QualType) {}
+  virtual void handlePointerType(const FieldDecl *, QualType) {}
+  virtual void handleArrayType(const FieldDecl *, QualType) {}
+  virtual void handleScalarType(const FieldDecl *, QualType) {}
+  // Most handlers shouldn't be handling this, just the field checker.
+  virtual void handleOtherType(const FieldDecl *, QualType) {}
+
+  // The following are only used for keeping track of where we are in the base
+  // class/field graph. Int Headers use this to calculate offset, most others
+  // don't have a need for these.
+
+  virtual void enterStruct(const CXXRecordDecl *, const FieldDecl *) {}
```

Comment:
I don't know if leaveStruct will ever be needed. The offsets generated for the integration header are monotonically increasing. If the lambda object members are visited in ascending sequence, the offset value should be steadily increasing, thus it will never need bumping down.

I confess I don't fully understand the base class stuff either. The old code made no reference to base classes.

Reply:
I did the 'leaveStruct' so that we can 'undo' the offset, that way we don't have to store the entire 'stack' of parents.  Instead, we can store just a single offset, and add/subtract it as needed.  For example, if we are deep in a set of structures and see:

    struct S {  accessor a;    };
    struct U { /*other fields;*/S s1, s2; };

We would have to 'know' about U and the entire set of parents in order to calculate the offset.  By doing the 'leaveStruct' (see the Int Header implementation), each time you enter a field, you simply increase by your offset, then decrease by the same offset when we leave.  So in the above:

We're in 'U', offset is something like 95 (random number).
We enter 's1', which we know is at offset '40' in 'U', so we increase our offset by that much.
We leave 's1', we can calculate the offset as '40' in 'U' and just decrease it, getting us back to 95.
Then, we enter 's2', we can do the same.

Otherwise, we'd have to know 'U''s offset, and our current field/bases' offset as well.  I couldn't come up with a way of doing that which didn't store the stack of currently visited things.